File: BaseLua/src/main/java/org/anime_game_servers/lua/engine/LuaScript.java
package org.anime_game_servers.lua.engine;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.script.ScriptException;
import java.util.List;
import java.util.Map;

public interface LuaScript {
    boolean hasMethod(@Nonnull String methodName);

    @Nullable
    LuaValue callMethod(@Nonnull String methodName, Object... args) throws ScriptException, NoSuchMethodException;

    void evaluate() throws ScriptException;

    <T> T getGlobalVariable(String name, Class<T> type);

    <T> List<T> getGlobalVariableList(String name, Class<T> type);

    <T> Map<String, T> getGlobalVariableMap(String name, Class<T> type);

    LuaEngine getEngine();
}

File: BaseLua/src/main/java/org/anime_game_servers/lua/engine/LuaScript.kt
package org.anime_game_servers.lua.engine

import javax.annotation.Nonnull
import javax.script.ScriptException

interface LuaScript {
    fun hasMethod(@Nonnull methodName: String): Boolean

    @Throws(ScriptException::class, NoSuchMethodException::class)
    fun callMethod(@Nonnull methodName: String, vararg args: Any?): LuaValue?

    @Throws(ScriptException::class)
    fun evaluate()
    fun <T> getGlobalVariable(name: String, type: Class<T>): T?
    fun <T> getGlobalVariableList(name: String, type: Class<T>): List<T>
    fun <T> getGlobalVariableMap(name: String, type: Class<T>): Map<String, T>
    val engine: LuaEngine?
}
